### PR TITLE
add amp on bert

### PR DESCRIPTION
**Pull request by xinyual**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

File: scripts/pretraining/bert/run_pretraining.py [hunk 1]
```diff
@@ -321,10 +340,17 @@ def train(args):
...
-
-            for loss in loss_l:
-                loss.backward()
-
+            if args.use_amp:
+                with mx.autograd.record():
+                    with amp.scale_loss(loss_l, trainer) as loss_l:
+                        for loss in loss_l:
+                            loss.backward()
+                norm_clip_mult = num_workers * trainer._amp_loss_scaler.loss_scale
```
Comment (sxjscience) [inline scripts/pretraining/bert/run_pretraining.py:348]: We can directly use `trainer.amp_loss_scale` as in https://github.com/dmlc/gluon-nlp/blob/dfdfb6d12f82e6caa47bbd7430efcd64dfa311dc/scripts/machine_translation/train_transformer.py#L708

**Comment by leezu** [comment]
Please resolve the conflicts

File: scripts/pretraining/bert/create_pretraining_data.py [hunk 2]
```diff
@@ -538,3 +538,171 @@ def truncate_seq_pair(tokens_a, tokens_b, max_num_tokens):
...
+def main():
+    """Main function."""
+    time_start = time.time()
+
+    # random seed
+    random.seed(args.random_seed)
+
+    # create output dir
+    output_dir = os.path.expanduser(args.output_dir)
+    # nlp.utils.mkdir(output_dir)
```
Comment (leezu) [inline scripts/pretraining/bert/create_pretraining_data.py:551]: Should this be uncommented or deleted?

File: scripts/pretraining/bert/README.md [hunk 3]
```diff
@@ -51,4 +108,5 @@ gluon_en_cased_bert_base
 ├── model-{short_hash}.yml    
 ```
 
-The pretrained model has been finetuned on SQUAD 2.0, obtaining F1/EM as 78.50/75.57. See also [question_answering](../../question_answering) for more details on the finetuning results.
+The pretrained model has been finetuned on SQUAD 2.0, obtaining F1/EM as 77.09/74.38. See also [question_answering](../../question_answering) for more details on the finetuning results.
+And when using AMP, we can refer this sheet https://docs.google.com/spreadsheets/d/1c5QPLBpJEAVDUk_BmcpKphp8nN3Dk9q2bKYqa3AIUz0/edit?usp=sharing to see more training details in phase1 and we get F1/EM 74.20/76.87.
```
Comment (sxjscience) [inline scripts/pretraining/bert/README.md:112]: Should be `F1/EM = 76.87/74.20`

File: scripts/pretraining/bert/README.md [hunk 4]
```diff
@@ -51,4 +108,5 @@ gluon_en_cased_bert_base
 ├── model-{short_hash}.yml    
 ```
 
-The pretrained model has been finetuned on SQUAD 2.0, obtaining F1/EM as 78.50/75.57. See also [question_answering](../../question_answering) for more details on the finetuning results.
+The pretrained model has been finetuned on SQUAD 2.0, obtaining F1/EM as 77.09/74.38. See also [question_answering](../../question_answering) for more details on the finetuning results.
```
Comment (sxjscience) [inline scripts/pretraining/bert/README.md:111]: It seems that the performance becomes worse.

**Review by leezu — APPROVED** [review]

**Pull request merged**